File: Springboot_Implementation/groceryapp/src/main/java/com/example/groceryapp/service/StoreService.java
package com.example.groceryapp.service;

import com.example.groceryapp.entity.Store;
import com.example.groceryapp.repository.StoreRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class StoreService {

    @Autowired
    private StoreRepository storeRepository;

    // CREATE
    public Store createStore(Store store) {
        return storeRepository.save(store);
    }

    // READ ALL
    public List<Store> getAllStores() {
        return storeRepository.findAll();
    }

    // READ BY ID
    public Optional<Store> getStoreById(Integer id) {
        return storeRepository.findById(id);
    }

    // UPDATE
    public Store updateStore(Integer id, Store storeDetails) {

        Store store = storeRepository.findById(id).orElse(null);

        if (store != null) {
            store.setStoreName(storeDetails.getStoreName());
            store.setStoreAddress(storeDetails.getStoreAddress());
            store.setStoreContact(storeDetails.getStoreContact());

            return storeRepository.save(store);
        }

        return null;
    }

    // DELETE
    public void deleteStore(Integer id) {
        storeRepository.deleteById(id);
    }

    // SEARCH BY NAME
    public List<Store> searchByName(String name) {
        return storeRepository.findByStoreNameContaining(name);
    }

    // SEARCH BY LOCATION (ADDRESS)
    public List<Store> searchByLocation(String location) {
        return storeRepository.findByStoreAddressContaining(location);
    }
}

File: React_Implementation/grocery-app/frontend/src/services/storeService.js
import api, { firstSuccess } from './api';

export const storeService = {
  getAllStores: async () => (await firstSuccess([
    () => api.get('/api/stores'),
    () => api.get('/stores'),
  ])).data,
  getStoreById: async (id) => (await firstSuccess([
    () => api.get(`/api/stores/${id}`),
    () => api.get(`/stores/${id}`),
  ])).data,
  createStore: async (store) => (await firstSuccess([
    () => api.post('/api/stores', store),
    () => api.post('/stores', store),
  ])).data,
  updateStore: async (id, store) => (await firstSuccess([
    () => api.put(`/api/stores/${id}`, store),
    () => api.put(`/stores/${id}`, store),
  ])).data,
  deleteStore: async (id) => (await firstSuccess([
    () => api.delete(`/api/stores/${id}`),
    () => api.delete(`/stores/${id}`),
  ])).data,
};
